daccs_client/client.py

```diff
@@ -31,94 +25,96 @@ def __init__(self, fallback: Optional[bool] = True) -> None:
+            if self._fallback:
                 warnings.warn("Cannot retrieve cloud registry. Falling back to cached version")
-
-        if not self._using_fallback:
-            self._load_registry_from_cloud()
+                self._load_registry_from_cache()
+            else:
+                raise
         else:
-            self._load_registry_from_cache()
+            self._load_registry_from_cloud(registry)
 
         for node, node_details in self._registry.items():
             self._nodes[node] = DACCSNode(node, node_details)
 
     @property
-    def nodes(self):
+    def nodes(self) -> dict[str, DACCSNode]:
         return self._nodes
 
-    def __getitem__(self, node):
+    def __getitem__(self, node: str) -> DACCSNode:
         try:
             return self.nodes[node]
         except KeyError:
             raise UnknownNodeError(f"No node named '{node}' in the DACCS network.") from None
 
-    def _load_registry_from_cloud(self):
+    def _load_registry_from_cloud(self, registry_response: requests.Response) -> None:
         try:
-            self._registry = self._registry.json()
-        except json.json.JSONDecodeError as e:
+            self._registry = registry_response.json()
+        except json.JSONDecodeError:
             raise RuntimeError(
                 "Could not parse JSON returned from the cloud registry. "
-                "Consider re-trying with 'fallback' set to True when instantiating the DACCS object."
+                f"Consider re-trying with 'fallback' set to True when instantiating the {self.__class__.__name__} "
+                "object."
             )
         self._save_registry_as_cache()
 
     def _load_registry_from_cache(self):
         try:
-            with open(CACHE, "r") as f:
+            with open(CACHE_FNAME, "r") as f:
                 self._registry = json.load(f)
-        except FileNotFoundError as e:
-            raise RuntimeError(f"Local registry cache not found. No file named {CACHE}.") from None
+        except FileNotFoundError:
+            raise RuntimeError(f"Local registry cache not found. No file named {CACHE_FNAME}.") from None
 
         try:
-            with open(CACHE_META, "r") as f:
+            with open(CACHE_META_FNAME, "r") as f:
                 data = json.load(f)
                 date = dateutil.parser.isoparse(data["last_cache_date"])
-        except (FileNotFoundError, ValueError) as e:
+        except (FileNotFoundError, ValueError):
             date = "Unknown"
 
         print(f"Registry loaded from cache dating: {date}")
 
     def _save_registry_as_cache(self):
-        cache_backup = CACHE + ".backup"
-        cache_meta_backup = CACHE_META + ".backup"
+        cache_backup = CACHE_FNAME + ".backup"
+        cache_meta_backup = CACHE_META_FNAME + ".backup"
+
+        # Create cache parent directories if they don't exist
+        for cache_dir in {os.path.dirname(CACHE_FNAME), os.path.dirname(CACHE_META_FNAME)}:
```

Comment:
Now that the cache files are in the user cache directory by default (or anywhere else if the user wants), we need to make sure that the files can actually be written to the specified location.